### PR TITLE
F/WPC-87 & F/WPC-55 comment frontend functionality

### PR DESCRIPTION
**Pull request by 0815pascal**
- Allows a reader (no matter if logged in or not) to read comments belonging to an article
- Allows a registered user to write comments
- Allows a registered user to reply to other comments
- Notifies a registered user whether his comment is waiting approval or needs adaptations (has been rejected)

### CODE REVIEW

**Review by roosnic1 — COMMENTED** [review]
looks good.

**Review by roosnic1 — CHANGES_REQUESTED** [review]
@0815pascal please clean up the PR first.

File: examples/website/src/shared/atoms/addBlockInput.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      <Dropdown
+        disabled={disabled}
+        renderTitle={() => {
+          return <IconButton appearance="primary" icon={<Icon icon="plus" />} circle />
+        }}>
+        {menuItems.map((item, index) => (
+          <Dropdown.Item
+            key={index}
+            onSelect={event => {
+              onMenuItemClick(item)
```
Comment (penk85) [inline examples/website/src/shared/atoms/addBlockInput.tsx:42]: Is event necessary here?

File: examples/website/package.json [hunk 2]
```diff
@@ -41,7 +41,9 @@
     "react-apollo": "^3.1.3",
     "react-dom": "^16.9.0",
     "react-helmet-async": "^1.0.2",
-    "slate": "^0.52.5"
+    "rsuite": "^4.8.4",
```
Comment (roosnic1) [inline examples/website/package.json:44]: Why did you add rsuite?

File: examples/website/src/shared/atoms/addBlockInput.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import React from 'react'
```
Comment (roosnic1) [inline examples/website/src/shared/atoms/addBlockInput.tsx:1]: why is this file in the PR?

File: examples/website/src/shared/atoms/blockList.tsx [hunk 4]
```diff
@@ -0,0 +1,307 @@
+import React, {Fragment, useState, ReactNode, useCallback, useMemo, memo} from 'react'
```
Comment (roosnic1) [inline examples/website/src/shared/atoms/blockList.tsx:1]: why is this file in the PR?

File: examples/website/src/shared/molecules/relatedPage.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
+import React from 'react'
```
Comment (roosnic1) [inline examples/website/src/shared/molecules/relatedPage.tsx:1]: What is this?

File: examples/website/src/shared/navigation/pageFooter.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
+import React from 'react'
```
Comment (roosnic1) [inline examples/website/src/shared/navigation/pageFooter.tsx:1]: why is this file completely new?

**Review by roosnic1 — CHANGES_REQUESTED** [review]
Needs more cleanup

File: examples/website/src/shared/atoms/blockList.tsx (hunk 4 above)
Comment (bishoymelek) [reply examples/website/src/shared/atoms/blockList.tsx:1]: @roosnic1 I guess because it's used in RichText Block

File: examples/website/src/shared/blocks/blockRenderer.tsx [hunk 7]
```diff
@@ -57,7 +57,7 @@ export function renderBlock(block: Block | null, opts: RenderBlockOptions) {
 
   switch (block.type) {
     case BlockType.RichText:
-      return <RichTextBlock value={block.value} />
+      return <RichTextBlock value={block.value} onChange={console.log} />
```
Comment (penk85) [inline examples/website/src/shared/blocks/blockRenderer.tsx:60]: Do you need the log here?

**Review by bishoymelek — CHANGES_REQUESTED** [review]
Great work, some tiny points

File: examples/website/src/shared/route/articleAdapter.ts [hunk 8]
```diff
@@ -26,6 +27,24 @@ export function authorsAdapter(authors: any): Author[] {
   })
 }
 
+export function commentsAdapter(comments: any): Comment[] {
```
Comment (penk85) [inline examples/website/src/shared/route/articleAdapter.ts:30]: Should this be typed more specifically?

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+            <h4 className={css(CommentAuthor)}>{userName}</h4>
+            <p className={css(CommentMeta)}>
+              <span>{authorType}</span> ·{' '}
+              <span className={css(Timestamp)}>{new Date(modifiedAt).toLocaleString()}</span>
+            </p>
+          </div>
+        </div>
+
+        <div className={css(CommentBody)}>
+          <RichTextBlock disabled value={text} onChange={value => value} />
```
Comment (roosnic1) [inline examples/website/src/shared/molecules/commentTemplate.tsx:310]: What is happening inside `onChange`? If you want to make a function do nothing at all you can write `() => { /* do nothing*/ }`

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+  readonly itemType: string
+  readonly itemID?: string
+  readonly slug?: string
+}
+
+// Queries
+// =======
+
+const AddComment = gql`
+  mutation AddComment($input: CommentInput!) {
```
[the remark this answers is not in the document]
Comment (roosnic1) [reply examples/website/src/shared/molecules/commentTemplate.tsx:176]: maybe move this graphQL code to `query.tsx`

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+
+  const {id, rejectionReason, state, user, authorType, modifiedAt, text} = value.value
+
+  return (
+    <div className={css(CommentBox, Replies)} id={id}>
+      <a href={'#' + id} className={css(commentBorderLink)}>
+        <span className={css(SrOnly)}>Jump to comment {id}</span>
+      </a>
+      <div className={css(CommentHeading)}>
+        <div style={{width: 50}}>
```
Comment (roosnic1) [inline examples/website/src/shared/molecules/commentTemplate.tsx:364]: We don't have images yet.

File: examples/website/src/shared/route/authorTemplateContainer.tsx [hunk 12]
```diff
@@ -131,7 +131,7 @@ export function AuthorTemplateContainer({id}: AuthorProps) {
       )}
 
       {/* {bio && <RichText value={bio} />} */}
-      {bio && <RichTextBlock value={bio} />}
+      {bio && <RichTextBlock value={bio} onChange={console.log} />}
```
Comment (roosnic1) [inline examples/website/src/shared/route/authorTemplateContainer.tsx:134]: `() => { /* do nothing */ }`

File: examples/website/src/shared/route/gqlFragments.ts [hunk 13]
```diff
@@ -110,6 +143,11 @@ export const pageMetaDataFragment = gql`
     image {
       ...SimpleImageData
     }
+    # Not yet needed in MVP
```
Comment (roosnic1) [inline examples/website/src/shared/route/gqlFragments.ts:146]: If not there please remove.

File: packages/api/src/graphql/mutation.private.ts [hunk 14]
```diff
@@ -725,10 +725,16 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
       async resolve(root, {id}, {dbAdapter, loaders}) {
         const article = await loaders.articles.load(id)
 
-        if(!article) throw new NotFound('article', id)
+        if (!article) throw new NotFound('article', id)
```
Comment (roosnic1) [inline packages/api/src/graphql/mutation.private.ts:728]: Did you do any functional changes in this file? If not please don't add it to the PR.

**Review by roosnic1 — CHANGES_REQUESTED** [review]

File: examples/website/src/shared/blocks/listicalBlock.tsx [hunk 15]
```diff
@@ -3,7 +3,7 @@ import {cssRule, useStyle} from '@karma.run/react'
 import {Color} from '../style/colors'
 import {pxToRem, whenTablet, whenDesktop} from '../style/helpers'
 import {Node} from 'slate'
-import {RichText} from '../atoms/richText'
+import {RichTextBlock} from './richTextBlock/richTextBlock'
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/blocks/listicalBlock.tsx:6]: It was my understanding that `RichText` was a "legacy version" of `RichTextBlock` so therefore I replaced it. But of course I can be wrong with my interpretation?

File: examples/website/src/shared/blocks/listicalBlock.tsx [hunk 16]
```diff
@@ -142,7 +142,7 @@ export function ListicalBLock({listical}: ListicalBLockProps) {
             )}
             {item.text && (
               <div className={css(ListicalTextStyle)}>
-                <RichText value={item.text} />
+                <RichTextBlock value={item.text} onChange={console.log} />
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/blocks/listicalBlock.tsx:145]: Done

File: examples/website/src/shared/blocks/peerPageBreak.tsx [hunk 17]
```diff
@@ -111,7 +111,7 @@ export function PageBreakBlock({
           />
         )}
       </div>
-      {!!richText && <RichText value={richText} />}
+      {!!richText && <RichTextBlock value={richText} onChange={console.log} />}
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/blocks/peerPageBreak.tsx:114]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 18]
```diff
@@ -0,0 +1,427 @@
...
+    },
+    onError: error => {
+      console.error('Error creating a post', error)
+      setCommentState("Something went wrong, your comment couldn't be saved")
+    }
+  })
+
+  return (
+    <div className={role === 'reply' ? css(ReplyBox) : css(Container)}>
+      {header ? <h3>{header}</h3> : ''}
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:210]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 19]
```diff
@@ -0,0 +1,427 @@
...
+        {commentState && <p className={css(StateMessage)}>{commentState}</p>}
+        <div className={css(CommentInput)}>
+          <RichTextBlock
+            value={commentInput}
+            onChange={input => {
+              setCommentInput(input)
+            }}
+          />
+        </div>
+        <p>
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:234]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx (hunk 9 above)
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:310]: done

File: examples/website/src/shared/route/authorTemplateContainer.tsx (hunk 12 above)
Comment (0815pascal) [reply examples/website/src/shared/route/authorTemplateContainer.tsx:134]: done

File: examples/website/src/shared/route/gqlFragments.ts (hunk 13 above)
Comment (0815pascal) [reply examples/website/src/shared/route/gqlFragments.ts:146]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 20]
```diff
@@ -0,0 +1,427 @@
...
+            disabled={loading}>
+            {props.parentCommentAuthor ? 'Submit' : 'Post comment'}
+          </BaseButton>
+        </p>
+      </form>
+    </div>
+  )
+}
+
+export function DisplayComments(props: DisplayCommentsProps) {
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:246]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 21]
```diff
@@ -0,0 +1,427 @@
...
+  const [commentID, setCommentID] = useState('')
+  const {comments} = props
+
+  return (
+    <div className={css(Container, CommentBox)}>
+      <h3 style={{marginTop: '2em'}}>Showing all comments</h3>
+      {comments?.map(parentComment => (
+        <ParentComment
+          comment={parentComment}
+          key={parentComment.id}
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:258]: Yes it seems like I need it; when setting it in the root tag of the component I would get the typical error-message.

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 22]
```diff
@@ -0,0 +1,427 @@
...
+  const {activeComment} = props
+
+  return (
+    <div className={css(Thread)}>
+      <div className={css(CommentBox)} id={id}>
+        <a href={'#' + id} className={css(commentBorderLink)}>
+          <span className={css(SrOnly)}>Jump to comment {id}</span>
+        </a>
+        <div className={css(CommentHeading)}>
+          <div style={{width: 50}}>
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:297]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx (hunk 11 above)
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:364]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 23]
```diff
@@ -0,0 +1,427 @@
...
+          </div>
+        </div>
+
+        <div className={css(CommentBody)}>
+          <RichTextBlock disabled value={text} onChange={value => value} />
+          <div className={css(Actions)}>
+            {state === 'approved' ? (
+              <button
+                className={css(SmallButton)}
+                onClick={() => props.handleCurrentComment(id === activeComment ? '' : id)}>
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:315]: Done

File: examples/website/src/shared/blocks/listicalBlock.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/blocks/listicalBlock.tsx:145]: Done

File: examples/website/src/shared/blocks/blockRenderer.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/blocks/blockRenderer.tsx:60]: Done

File: examples/website/src/shared/atoms/primaryButton.tsx [hunk 24]
```diff
@@ -13,13 +13,13 @@ const PrimaryButtonStyle = cssRule({
   border: `1px solid ${Color.NeutralDark}`,
 
   fontSize: '1em',
-  borderRadius: '3em',
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/atoms/primaryButton.tsx:16]: It's something that was only used once and I found `3em` to be in general a bit over the top. Buttons in this shape remind me of the time when I first laid my hands on WebDev in 2002 😄

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 25]
```diff
@@ -0,0 +1,427 @@
...
+        </p>
+      </form>
+    </div>
+  )
+}
+
+export function DisplayComments(props: DisplayCommentsProps) {
+  const css = useStyle()
+
+  const [commentID, setCommentID] = useState('')
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:249]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 26]
```diff
@@ -0,0 +1,427 @@
...
+            setCommentID(commentID)
+          }
+          activeComment={commentID}
+        />
+      ))}
+    </div>
+  )
+}
+
+function ParentComment(props: any) {
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:269]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx [hunk 27]
```diff
@@ -0,0 +1,427 @@
...
+export function DisplayComments(props: DisplayCommentsProps) {
+  const css = useStyle()
+
+  const [commentID, setCommentID] = useState('')
+  const {comments} = props
+
+  return (
+    <div className={css(Container, CommentBox)}>
+      <h3 style={{marginTop: '2em'}}>Showing all comments</h3>
+      {comments?.map(parentComment => (
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:255]: Done

File: examples/website/src/shared/molecules/commentTemplate.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/molecules/commentTemplate.tsx:176]: Done

File: examples/website/src/shared/route/articleAdapter.ts (hunk 8 above)
Comment (0815pascal) [reply examples/website/src/shared/route/articleAdapter.ts:30]: Good question, I don't know. I followed the examples from the other Adapters in the file (`authorsAdapter`, `articleMetaAdapter`, etc.)  and they also use `any` at this place. What do you think @roosnic1 ?

File: packages/api/src/graphql/mutation.private.ts (hunk 14 above)
Comment (0815pascal) [reply packages/api/src/graphql/mutation.private.ts:728]: Tried solving this problem but to no avail. Actually I figured that it might not even be a problem in that sense because if you look in this file at `publishPage` for example there's also a space between the `if` and the condition. So prettier actually only does what it's supposed to do (in my understanding at least). But if you'd like I can try to edit this manually from within Github?

File: examples/website/src/shared/route/articleAdapter.ts (hunk 8 above)
Comment (roosnic1) [reply examples/website/src/shared/route/articleAdapter.ts:30]: It's always better to have more types but since this will likely change in the future anyway we can leave it at any,

File: examples/website/src/shared/types.ts [hunk 28]
```diff
@@ -1,10 +1,5 @@
 import {Node} from 'slate'
-import {
-  CommentAuthorType,
-  CommentItemType,
-  RichTextTextNode,
-  User
-} from '../../../../packages/api/lib'
+import {CommentAuthorType, CommentItemType, User} from '../../../../packages/api/lib'
```
Comment (roosnic1) [inline examples/website/src/shared/types.ts:2]: This import looks very wrong? Did you add this?

**Review by roosnic1 — CHANGES_REQUESTED** [review]

File: examples/website/src/shared/route/authorTemplateContainer.tsx [hunk 29]
```diff
@@ -131,7 +131,14 @@ export function AuthorTemplateContainer({id}: AuthorProps) {
       )}
 
       {/* {bio && <RichText value={bio} />} */}
-      {bio && <RichTextBlock value={bio} />}
+      {bio && (
```
Comment (bishoymelek) [inline examples/website/src/shared/route/authorTemplateContainer.tsx:134]: we have to add displayOnly here?

**Review by bishoymelek — CHANGES_REQUESTED** [review]

File: examples/website/src/shared/route/articleAdapter.ts [hunk 30]
```diff
@@ -1,4 +1,5 @@
-import {PublishedArticle, Author, ArticleMeta, Peer} from '../types'
+// import { User } from '../../../../../packages/api/lib/db/user'
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/route/articleAdapter.ts:1]: We will need it later for the RichTextBlock when the issue with Server-Side Rendering of *.less-files is resolved

File: examples/website/src/shared/route/articleTemplateContainer.tsx [hunk 31]
```diff
@@ -389,7 +397,7 @@ export function PeerProfileBlock({peer, article}: PeerProfileBlockProps) {
       <div className={css(PeerProfileCallToActionURL)}>
         {peer?.callToActionText?.length && (
           <a target="_blank" rel="noreferrer" href={peer?.callToActionURL}>
-            <RichText value={peer?.callToActionText} />
+            <RichTextBlock value={peer?.callToActionText} disabled onChange={console.log} />
```
[the remark this answers is not in the document]
Comment (0815pascal) [reply examples/website/src/shared/route/articleTemplateContainer.tsx:400]: Done

File: examples/website/src/shared/route/authorTemplateContainer.tsx (hunk 29 above)
Comment (0815pascal) [reply examples/website/src/shared/route/authorTemplateContainer.tsx:134]: Done

File: examples/website/src/shared/types.ts (hunk 28 above)
Comment (0815pascal) [reply examples/website/src/shared/types.ts]: Done

**Review by roosnic1 — APPROVED** [review]
👍🏾

**Review by bishoymelek — APPROVED** [review]
:tada:
